### PR TITLE
feat(backingimage): support spdk backing image

### PR DESCRIPTION
**Pull request by ChanYiLin**
ref: https://github.com/longhorn/longhorn/issues/6341

- Add BackingImage Proxy API to BackingImage SPDK Server API

### CODE REVIEW

File: pkg/meta/version.go [hunk 1]
```diff
@@ -12,6 +12,10 @@ const (
 	// InstanceManagerDiskServiceAPIVersion used to communicate with the user e.g. longhorn-manager
 	InstanceManagerDiskServiceAPIVersion    = 1
 	InstanceManagerDiskServiceAPIMinVersion = 1
+
+	// InstanceManagerBackingImageServiceAPIVersion is used for compatibility check for longhorn-manager
+	InstanceManagerBackingImageServiceAPIVersion    = 1
+	InstanceManagerBackingImageServiceAPIMinVersion = 1
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply pkg/meta/version.go:18]: No, we don't this anymore as we move the API to proxy server.

File: pkg/api/backing_image.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		ExpectedChecksum: obj.Spec.Checksum,
+
+		Status: BackingImageStatus{
+			Progress:        int(obj.Status.Progress),
+			State:           obj.Status.State,
+			CurrentChecksum: obj.Status.Checksum,
+			ErrorMsg:        obj.Status.ErrorMsg,
+		},
+	}
+}
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply pkg/api/backing_image.go:40]: fixed.

File: pkg/proxy/backing_image.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+func (p *Proxy) SPDKBackingImageGet(ctx context.Context, req *rpc.SPDKBackingImageGetRequest) (*rpc.SPDKBackingImageResponse, error) {
+	log := logrus.WithFields(logrus.Fields{
+		"name":     req.Name,
+		"diskUuid": req.DiskUuid,
+	})
+	log.Info("Backing Image Server: Get SPDk Backing Image")
+	if req.Name == "" || req.DiskUuid == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "failed to create spdk backing image due to parameters missing")
+	}
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply pkg/proxy/backing_image.go:71]: Fixed.

File: pkg/proxy/backing_image.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+func (p *Proxy) SPDKBackingImageDelete(ctx context.Context, req *rpc.SPDKBackingImageDeleteRequest) (*emptypb.Empty, error) {
+	log := logrus.WithFields(logrus.Fields{
+		"name":     req.Name,
+		"diskUuid": req.DiskUuid,
+	})
+	log.Info("Backing Image Server: Deleting SPDk Backing Image")
+	if req.Name == "" || req.DiskUuid == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "failed to create spdk backing image due to parameters missing")
+	}
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply pkg/proxy/backing_image.go:59]: Fixed.

File: pkg/proxy/backing_image.go [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+				return errors.Wrap(err, "failed to send backing image response")
+			}
+		}
+	}
+}
+
+func (p *Proxy) watchSPDKBackingImage(ctx context.Context, req *emptypb.Empty, client *spdkclient.SPDKClient, notifyChan chan struct{}) error {
+	logrus.Info("Start watching SPDK replicas")
+
+	notifier, err := client.BackingImageWatch(context.Background())
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply pkg/proxy/backing_image.go:162]: fixed.

File: pkg/proxy/backing_image.go [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+func (p *Proxy) SPDKBackingImageWatch(req *emptypb.Empty, srv rpc.ProxyEngineService_SPDKBackingImageWatchServer) error {
+	logrus.Info("Start watching SPDK backing image")
+
+	done := make(chan struct{})
+
+	// Create a client for watching SPDK backing image
+	spdkClient, err := spdkclient.NewSPDKClient(p.spdkServiceAddress)
+	if err != nil {
+		done <- struct{}{}
+		return grpcstatus.Error(grpccodes.Internal, errors.Wrapf(err, "failed to create SPDK client").Error())
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply pkg/proxy/backing_image.go:104]: fixed.

File: pkg/meta/version.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply pkg/meta/version.go:18]: removed

File: pkg/api/backing_image.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
+package api
+
+import (
+	rpc "github.com/longhorn/types/pkg/generated/imrpc"
+	"google.golang.org/protobuf/types/known/emptypb"
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply pkg/api/backing_image.go:5]: fixed.

File: pkg/client/proxy_backing_image.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+	"github.com/longhorn/longhorn-instance-manager/pkg/types"
+	rpc "github.com/longhorn/types/pkg/generated/imrpc"
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+func (c *ProxyClient) SPDKBackingImageCreate(name, backingImageUUID, diskUUID, checksum, fromAddress, srcLvsUUID string, size uint64) (*api.BackingImage, error) {
+	if name == "" || backingImageUUID == "" || checksum == "" || diskUUID == "" || size == 0 {
+		return nil, fmt.Errorf("failed to create backing image: missing required parameters")
+	}
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply pkg/client/proxy_backing_image.go:17]: fixed.

File: pkg/client/proxy_backing_image.go [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+		Name:     name,
+		DiskUuid: diskUUID,
+	})
+	return err
+}
+
+func (c *ProxyClient) SPDKBackingImageGet(name, diskUUID string) (*api.BackingImage, error) {
+	if name == "" || diskUUID == "" {
+		return nil, fmt.Errorf("failed to get backing image: missing required parameters")
+	}
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply pkg/client/proxy_backing_image.go:58]: fixed.

File: pkg/proxy/proxy.go [hunk 10]
```diff
@@ -6,8 +6,11 @@ import (
 
 	eclient "github.com/longhorn/longhorn-engine/pkg/controller/client"
 	"github.com/longhorn/types/pkg/generated/enginerpc"
+	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 	"golang.org/x/net/context"
+	grpccodes "google.golang.org/grpc/codes"
+	grpcstatus "google.golang.org/grpc/status"
 	"google.golang.org/protobuf/types/known/emptypb"
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply pkg/proxy/proxy.go:15]: fixed.

File: pkg/client/proxy_backing_image.go [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/longhorn/longhorn-instance-manager/pkg/api"
+	"github.com/longhorn/longhorn-instance-manager/pkg/types"
+	rpc "github.com/longhorn/types/pkg/generated/imrpc"
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/types/known/emptypb"
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply pkg/client/proxy_backing_image.go:11]: fixed.

File: pkg/proxy/backing_image.go [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+
+func (p *Proxy) SPDKBackingImageDelete(ctx context.Context, req *rpc.SPDKBackingImageDeleteRequest) (*emptypb.Empty, error) {
+	log := logrus.WithFields(logrus.Fields{
+		"name":     req.Name,
+		"diskUuid": req.DiskUuid,
+	})
+	log.Info("Backing Image Server: Deleting SPDk Backing Image")
+	if req.Name == "" || req.DiskUuid == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "failed to delete spdk backing image due to parameters missing")
+	}
```
Comment (derekbit) [inline pkg/proxy/backing_image.go:59]: Use `validateProxyMethodParameters`

e.g.

```
	input := map[string]string{
		"engineName":     engineName,
		"volumeName":     volumeName,
		"serviceAddress": serviceAddress,
	}
	if err := validateProxyMethodParameters(input); err != nil {
		return nil, errors.Wrap(err, "failed to list replicas for volume")
	}
```

File: pkg/proxy/backing_image.go [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+
+func (p *Proxy) SPDKBackingImageGet(ctx context.Context, req *rpc.SPDKBackingImageGetRequest) (*rpc.SPDKBackingImageResponse, error) {
+	log := logrus.WithFields(logrus.Fields{
+		"name":     req.Name,
+		"diskUuid": req.DiskUuid,
+	})
+	log.Info("Backing Image Server: Get SPDk Backing Image")
+	if req.Name == "" || req.DiskUuid == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "failed to get spdk backing image due to parameters missing")
+	}
```
Comment (derekbit) [inline pkg/proxy/backing_image.go:71]: Use `validateProxyMethodParameters`

File: pkg/proxy/backing_image.go [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+	return &emptypb.Empty{}, p.spdkLocalClient.BackingImageDelete(req.Name, req.DiskUuid)
+}
+
+func (p *Proxy) SPDKBackingImageGet(ctx context.Context, req *rpc.SPDKBackingImageGetRequest) (*rpc.SPDKBackingImageResponse, error) {
+	log := logrus.WithFields(logrus.Fields{
+		"name":     req.Name,
+		"diskUuid": req.DiskUuid,
+	})
+	log.Info("Backing Image Server: Get SPDk Backing Image")
```
Comment (derekbit) [inline pkg/proxy/backing_image.go:68]:
```suggestion
	log.Debug("Backing Image Server: Get SPDk Backing Image")
```

File: pkg/proxy/backing_image.go [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+	}
+	ret, err := p.spdkLocalClient.BackingImageGet(req.Name, req.DiskUuid)
+	if err != nil {
+		return nil, grpcstatus.Error(grpccodes.Internal, err.Error())
+	}
+	return spdkBackingImageToBackingImageResponse(ret), nil
+}
+
+func (p *Proxy) SPDKBackingImageList(ctx context.Context, req *emptypb.Empty) (*rpc.SPDKBackingImageListResponse, error) {
+	logrus.Info("Backing Image Server: List SPDk Backing Image")
```
Comment (derekbit) [inline pkg/proxy/backing_image.go:80]:
```suggestion
	logrus.Debug("Backing Image Server: List SPDk Backing Image")
```

File: pkg/proxy/backing_image.go [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+		"Checksum":         req.Checksum,
+		"FromAddress":      req.FromAddress,
+		"SrcLvsUuid":       req.SrcLvsUuid,
+	})
+
+	log.Info("Backing Image Server: Creating SPDk Backing Image")
+
+	if req.Name == "" || req.BackingImageUuid == "" || req.DiskUuid == "" || req.Checksum == "" || req.Size == 0 {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "failed to create spdk backing image due to parameters missing")
+	}
```
Comment (derekbit) [inline pkg/proxy/backing_image.go:42]: Use `validateProxyMethodParameters`

File: pkg/proxy/backing_image.go (hunk 12 above)
Comment (ChanYiLin) [reply pkg/proxy/backing_image.go:59]: removed the check

File: pkg/proxy/backing_image.go (hunk 13 above)
Comment (ChanYiLin) [reply pkg/proxy/backing_image.go:71]: removed

File: pkg/proxy/backing_image.go (hunk 14 above)
Comment (ChanYiLin) [reply pkg/proxy/backing_image.go:68]: fixed.

File: pkg/proxy/backing_image.go (hunk 15 above)
Comment (ChanYiLin) [reply pkg/proxy/backing_image.go:80]: fixed.

File: pkg/proxy/backing_image.go (hunk 16 above)
Comment (ChanYiLin) [reply pkg/proxy/backing_image.go:42]: removed.

**Review by derekbit — APPROVED** [review]

**Pull request merged**